tracecontext.go

```diff
@@ -169,23 +169,61 @@ type TraceState struct {
 
 // NewTraceState returns a TraceState based on entries.
 func NewTraceState(entries ...TraceStateEntry) TraceState {
-	out := TraceState{}
+	var out TraceState
 	var last *TraceStateEntry
+	var modified []TraceStateEntry
+	recorded := make(map[string]uint8)
+	// Range over the entries and find duplicate keys. When a previous key is
+	// found, it is stored in a map with a count that is incremented every time
+	// the key is observed.
+	// When keys are observed more than once, the last value observed is stored
+	// in a slice of the modified keys.
 	for _, e := range entries {
-		e := e // copy
+		if count, ok := recorded[e.Key]; ok {
+			if count > 1 {
+				for i, m := range modified {
+					if m.Key == e.Key {
+						modified[i] = e
+					}
+				}
+			} else {
+				modified = append(modified, e)
+			}
+		}
+		recorded[e.Key]++
+	}
+	// After we've ranged over the whole entry slice, any duplicates present
+	// must be written first in a last to first order, thus we range the slice
```

Comment:
nit: `must be written in a last to first order`, I think it makes the comment a bit simpler since we're already saying `last` and `first`